baseband/guppi/payload.py

```diff
@@ -0,0 +1,239 @@
+            ``cls._nbytes``, or, for mapping, to the end of the file).
+        **kwargs
+            Additional arguments are passed on to the class initializer. These
+            are only needed if ``header`` is not given.
+        """
+        if payload_nbytes is None:
+            payload_nbytes = (cls._nbytes if header is None
+                              else header.payload_nbytes)
+
+        if not memmap:
+            return super(GUPPIPayload, cls).fromfile(
+                fh, header=header, payload_nbytes=payload_nbytes, **kwargs)
+
+        if hasattr(fh, 'memmap'):
+            words = fh.memmap(dtype=cls._dtype_word,
+                              shape=None if payload_nbytes is None else
+                              (payload_nbytes // cls._dtype_word.itemsize,))
+        else:
+            mode = fh.mode.replace('b', '')
+            offset = fh.tell()
+            words = np.memmap(
+                fh, mode=mode, dtype=cls._dtype_word, offset=offset,
+                shape=(None if payload_nbytes is None
+                       else (payload_nbytes // cls._dtype_word.itemsize,)))
+            fh.seek(offset + words.size * words.dtype.itemsize)
+        return cls(words, header=header, **kwargs)
+
+    @classmethod
+    def fromdata(cls, data, header=None, bps=8, time_ordered=True):
+        """Encode data as a payload.
+
+        Parameters
+        ----------
+        data : `~numpy.ndarray`
+            Data to be encoded. The last dimension is taken as the number of
+            channels.
+        header : `~baseband.guppi.GUPPIHeader`, optional
+            If given, used to infer the ``bps`` and ``time_ordered``.
+        bps : int, optional
+            Bits per elementary sample, used if ``header`` is `None`.
+            Default: 8.
+        time_ordered : bool, optional
+            Whether data are time-ordered, used if ``header`` is `None`.
+            Default: `True`.
+        """
+        if header is not None:
+            bps = header.bps
+            time_ordered = header.time_ordered
+        sample_shape = data.shape[1:]
+        complex_data = data.dtype.kind == 'c'
+        try:
+            encoder = cls._encoders[bps]
+        except KeyError:
+            raise ValueError("{0} cannot encode data with {1} bits"
+                             .format(cls.__name__, bps))
+        # If time-ordered, switch to (nchan, nsample, npol); otherwise use
+        # (nsample, nchan, npol)
+        if time_ordered:
+            data = data.transpose(2, 0, 1)
+        else:
+            data = data.transpose(0, 2, 1)
+        if complex_data:
+            data = data.view((data.real.dtype, (2,)))
+        words = encoder(data).ravel().view(cls._dtype_word)
+        return cls(words, sample_shape=sample_shape, bps=bps,
+                   complex_data=complex_data, time_ordered=time_ordered)
+
+    def __len__(self):
+        """Number of samples in the payload."""
+        return self.nbytes * 8 // self._bpfs // (
```

Comment:
Here, we're undoing the `_bpfs` hack; how about keeping the "real" `_bpfs` as `_true_bpfs` in `__init__` and using that here?

Reply:
Added `_true_bpfs`.